### PR TITLE
fix(query) StitchRVS to accept RVRange for the entire user provided start/step/end

### PR DESCRIPTION
**Pull request by amolnayak311**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** 
   
Stitch creates an ``RvRange`` using a union of the children's ``RvRange``. This causes issues when Stitch is not the the top level operation and is under a ``BinaryJoinExec`` The lhs and rhs of Binary joins will not match causing inconsistent query results. 

**New behavior :**
``StitchRVSExec`` has been updated to now accept a ``RvRange`` that the user query expects. The union of ``RvRange`` therefore is not needed.

### CODE REVIEW

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 1]
```diff
@@ -270,6 +270,15 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
     (partitions, routingKeys)
   }
 
+  private def rvRangeFromLogicalPlan(logicalPlan: LogicalPlan, queryParams: PromQlQueryParams): Some[RvRange] = {
+    import LogicalPlan._
+    rvRangeFromPlan(logicalPlan) match {
+      case x @ Some(_)    =>  x
+      case None           => Some(RvRange(startMs = queryParams.startSecs * 1000L,
+        endMs = queryParams.endSecs * 1000L,
+        stepMs = queryParams.stepSecs * 1000L))
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:279]: Not understanding the case when we need to get start/step/end from queryParams. Can you explain the forcing factor here?

Comment (amolnayak311) [reply coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:279]: Since other parts of the code were using ``queryParams`` I have used the same values here.

File: query/src/main/scala/filodb/query/exec/StitchRvsExec.scala [hunk 2]
```diff
@@ -83,10 +84,7 @@ final case class StitchRvsExec(queryContext: QueryContext,
       groups.mapValues { toMerge =>
         val rows = StitchRvsExec.merge(toMerge.map(_.rows()))
         val key = toMerge.head.key
-        val outputRange = toMerge.map(_.outputRange).reduce { (rv1Range, rv2Range) =>
-          RvRange.union(rv1Range, rv2Range)
-        }
-        IteratorBackedRangeVector(key, rows, outputRange)
+        IteratorBackedRangeVector(key, rows, outputRvRange)
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/StitchRvsExec.scala:87]: The `union` function can be removed. It won't be used anymore.

File: query/src/main/scala/filodb/query/exec/StitchRvsExec.scala [hunk 3]
```diff
@@ -83,10 +84,7 @@ final case class StitchRvsExec(queryContext: QueryContext,
       groups.mapValues { toMerge =>
         val rows = StitchRvsExec.merge(toMerge.map(_.rows()))
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/StitchRvsExec.scala:85]: merge function now needs to accept outputRange and fill NaNs for absent rows.

**Review by vishramachandran — CHANGES_REQUESTED** [review]

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala (hunk 1 above)
Comment (alextheimer) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:279]: The `RangeVector::outputRange` spec says:
```
...Present only for time series data that is periodic. If raw data is requested, then None.
```
Does this break for an instant query like `foo[8d]`?

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 4]
```diff
@@ -270,6 +270,15 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
     (partitions, routingKeys)
   }
 
+  private def rvRangeFromLogicalPlan(logicalPlan: LogicalPlan, queryParams: PromQlQueryParams): Some[RvRange] = {
```
Comment (alextheimer) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:273]: See my other comment-- I think `rvRangeFromPlan` is sufficient.

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala [hunk 5]
```diff
@@ -195,10 +194,9 @@ class SingleClusterPlanner(val dataset: Dataset,
     rangeSelector match {
       case IntervalSelector(from, to) => TimeRangeChunkScan(from, to)
       case AllChunksSelector          => AllChunkScan
-      case EncodedChunksSelector      => ???
       case WriteBufferSelector        => WriteBufferChunkScan
       case InMemoryChunksSelector     => InMemoryChunkScan
-      case _                          => ???
+      case _                          => throw new IllegalArgumentException("Unsupported range selector found")
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala:199]: Include selector in error message

File: query/src/main/scala/filodb/query/exec/StitchRvsExec.scala [hunk 6]
```diff
@@ -12,41 +12,55 @@ import filodb.query.Query.qLogger
...
-  }
+  private class RVCursorImpl(vectors: Iterable[RangeVectorCursor], outputRange: Option[RvRange])
+    extends RangeVectorCursor {
+    private val bVectors = vectors.map(_.buffered)
+    val mins = new mutable.ArrayBuffer[BufferedIterator[RowReader]](2)
+    val nanResult = new TransientRow(0, Double.NaN)
+    val tsIter: BufferedIterator[Long] = outputRange match {
+      case Some(RvRange(startMs, stepMs, endMs)) =>
+        Iterator.iterate(startMs){_ + stepMs}.takeWhile( _ <= endMs).buffered
+      case None                                  => Iterator.iterate(Long.MaxValue)(_ =>Long.MaxValue).buffered
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/StitchRvsExec.scala:23]: Is there a reason not to error out if outputRange is None ?
I think we should, or else, bugs may silently creep in.

File: query/src/main/scala/filodb/query/exec/StitchRvsExec.scala [hunk 7]
```diff
@@ -12,41 +12,55 @@ import filodb.query.Query.qLogger
...
-            }
-            else if (t < minTime) {
-              mins.clear()
-              mins += r
-              minTime = t
-            } else if (t == minTime) {
-              mins += r
-            }
+    override def next(): RowReader = {
+
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/StitchRvsExec.scala:36]: Retain old comment:
```
// This is an n-way merge without using a heap.
// Heap is not used since n is expected to be very small (almost always just 1 or 2)
```

File: query/src/test/scala/filodb/query/exec/StitchRvsExecSpec.scala [hunk 8]
```diff
@@ -159,7 +159,65 @@ class StitchRvsExecSpec extends AnyFunSpec with Matchers with ScalaFutures {
...
+      .toListL.runAsync.futureValue
+    output.size shouldEqual 1
+
+    // Notice how the output RVRange is same as the one passed during initialization of the StitchRvsExec
+    output.head.outputRange shouldEqual Some(RvRange(0, 10, 150))
+    compareIter(output.head.rows().map(r => (r.getLong(0), r.getDouble(1))) , expected.toIterator)
+  }
+
+
+  it ("should merge RVs despite passing None for RvRange to StitchRvsExec") {
```
Comment (vishramachandran) [inline query/src/test/scala/filodb/query/exec/StitchRvsExecSpec.scala:201]: Per above comment, do we need to support this ?

File: query/src/test/scala/filodb/query/exec/StitchRvsExecSpec.scala [hunk 9]
```diff
@@ -159,7 +159,65 @@ class StitchRvsExecSpec extends AnyFunSpec with Matchers with ScalaFutures {
...
     val expected =
+      Seq(
+        (0L, Double.NaN),
+        (10L, 3d),
+        (20L, 3d),
+        (30L, 3d),
+        (40L, 3d),
+        (50L, 3d),
+        (60L, 3d),
+        (70L, 3d),
```
Comment (vishramachandran) [inline query/src/test/scala/filodb/query/exec/StitchRvsExecSpec.scala:170]: Can you add missing instants in input in the middle ?

File: query/src/test/scala/filodb/query/exec/StitchRvsExecSpec.scala [hunk 10]
```diff
@@ -297,48 +417,48 @@ class StitchRvsExecSpec extends AnyFunSpec with Matchers with ScalaFutures {
...
-      override def outputRange: Option[RvRange] = None
-    }
-
-    intercept[IllegalArgumentException] {
-      StitchRvsExec.stitch(r1, r2)
-    }
-
-  }
-
+// No more required as union is not invoked and we expect to see the RvRange passed by planner
```
Comment (vishramachandran) [inline query/src/test/scala/filodb/query/exec/StitchRvsExecSpec.scala:425]: Just reminder to remove commented code

**Review by vishramachandran — APPROVED** [review]
Thank you for the thorough work, and code cleanup.
Approved with some minor comments.

File: query/src/test/scala/filodb/query/exec/StitchRvsExecSpec.scala [hunk 11]
```diff
@@ -141,7 +141,7 @@ class StitchRvsExecSpec extends AnyFunSpec with Matchers with ScalaFutures {
     mergeAndValidate(rvs, expected)
   }
 
-  it ("should reduce output range") {
+  it ("should output range passed in to StitchRvsExec") {
```
Comment (vishramachandran) [inline query/src/test/scala/filodb/query/exec/StitchRvsExecSpec.scala:144]: This is the only positive test case I see. Any other edge cases that we need to validate ?
